loadbalancingexporter: Race in unit test

```
running make do-unit-tests-with-cover in ./exporter/loadbalancingexporter
make[2]: Entering directory '/home/circleci/project/exporter/loadbalancingexporter'
running go unit test ./... + coverage in /home/circleci/project/exporter/loadbalancingexporter
==================
WARNING: DATA RACE
Read at 0x00c00014a938 by goroutine 23:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeTraces()
      /home/circleci/project/exporter/loadbalancingexporter/exporter_test.go:643 +0x154e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1050 +0x1eb

Previous write at 0x00c00014a938 by goroutine 30:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).resolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:160 +0xb0b
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).periodicallyResolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:105 +0x265

Goroutine 23 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1095 +0x537
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1339 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1050 +0x1eb
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1337 +0x594
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1252 +0x2ff
  main.main()
      _testmain.go:188 +0x337

Goroutine 30 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).start()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:83 +0xf9
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*exporterImp).Start()
      /home/circleci/project/exporter/loadbalancingexporter/exporter.go:116 +0x164
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeTraces()
      /home/circleci/project/exporter/loadbalancingexporter/exporter_test.go:601 +0x11b2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1050 +0x1eb
==================
==================
WARNING: DATA RACE
Read at 0x00c0003281d0 by goroutine 23:
  reflect.typedmemmove()
      /usr/local/go/src/runtime/mbarrier.go:177 +0x0
  reflect.packEface()
      /usr/local/go/src/reflect/value.go:119 +0x126
  reflect.valueInterface()
      /usr/local/go/src/reflect/value.go:1030 +0x1b9
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:146 +0xc78
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:98 +0x10fe
  reflect.DeepEqual()
      /usr/local/go/src/reflect/deepequal.go:210 +0x339
  github.com/stretchr/testify/assert.ObjectsAreEqual()
      /home/circleci/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:65 +0x23b
  github.com/stretchr/testify/assert.Equal()
      /home/circleci/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:343 +0x257
  github.com/stretchr/testify/require.Equal()
      /home/circleci/go/pkg/mod/github.com/stretchr/testify@v1.7.0/require/require.go:162 +0xf8
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeTraces()
      /home/circleci/project/exporter/loadbalancingexporter/exporter_test.go:643 +0x15cb
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1050 +0x1eb

Previous write at 0x00c0003281d0 by goroutine 30:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).resolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:148 +0x972
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).periodicallyResolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:105 +0x265

Goroutine 23 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1095 +0x537
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1339 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1050 +0x1eb
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1337 +0x594
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1252 +0x2ff
  main.main()
      _testmain.go:188 +0x337

Goroutine 30 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).start()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:83 +0xf9
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*exporterImp).Start()
      /home/circleci/project/exporter/loadbalancingexporter/exporter.go:116 +0x164
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeTraces()
      /home/circleci/project/exporter/loadbalancingexporter/exporter_test.go:601 +0x11b2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1050 +0x1eb
==================
--- FAIL: TestRollingUpdatesWhenConsumeTraces (0.06s)
    testing.go:965: race detected during execution of test
FAIL
coverage: 97.9% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter	0.230s
FAIL
make[2]: *** [../../Makefile.Common:58: do-unit-tests-with-cover] Error 1
make[2]: Leaving directory '/home/circleci/project/exporter/loadbalancingexporter'
make[1]: *** [Makefile:69: for-all] Error 2
make[1]: Leaving directory '/home/circleci/project'
make: *** [Makefile:39: unit-tests-with-cover] Error 2
```

This might be related:

```
==================
WARNING: DATA RACE
Read at 0x00c000362258 by goroutine 69:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeLogs()
      /home/circleci/project/exporter/loadbalancingexporter/log_exporter_test.go:490 +0x1915
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Previous write at 0x00c000362258 by goroutine 70:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).resolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:160 +0xaa4
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).periodicallyResolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:105 +0x265

Goroutine 69 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1238 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1511 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1509 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1417 +0x3b3
  main.main()
      _testmain.go:221 +0x356

Goroutine 70 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).start()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:83 +0x104
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*loadBalancerImp).Start()
      /home/circleci/project/exporter/loadbalancingexporter/loadbalancer.go:104 +0x14f
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*logExporterImp).Start()
      /home/circleci/project/exporter/loadbalancingexporter/log_exporter.go:76 +0x99
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeLogs()
      /home/circleci/project/exporter/loadbalancingexporter/log_exporter_test.go:448 +0x1566
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000620090 by goroutine 69:
  reflect.typedmemmove()
      /usr/local/go/src/runtime/mbarrier.go:177 +0x0
  reflect.packEface()
      /usr/local/go/src/reflect/value.go:121 +0x12f
  reflect.valueInterface()
      /usr/local/go/src/reflect/value.go:1046 +0x1cd
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:154 +0x5bb
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:106 +0xe7e
  reflect.DeepEqual()
      /usr/local/go/src/reflect/deepequal.go:218 +0x349
  github.com/stretchr/testify/assert.ObjectsAreEqual()
      /home/circleci/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:65 +0x1e7
  github.com/stretchr/testify/assert.Equal()
      /home/circleci/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:343 +0x284
  github.com/stretchr/testify/require.Equal()
      /home/circleci/go/pkg/mod/github.com/stretchr/testify@v1.7.0/require/require.go:162 +0xf8
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeLogs()
      /home/circleci/project/exporter/loadbalancingexporter/log_exporter_test.go:490 +0x1992
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Previous write at 0x00c000620090 by goroutine 70:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).resolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:148 +0x913
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).periodicallyResolve()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:105 +0x265

Goroutine 69 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1238 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1511 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1509 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1417 +0x3b3
  main.main()
      _testmain.go:221 +0x356

Goroutine 70 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*dnsResolver).start()
      /home/circleci/project/exporter/loadbalancingexporter/resolver_dns.go:83 +0x104
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*loadBalancerImp).Start()
      /home/circleci/project/exporter/loadbalancingexporter/loadbalancer.go:104 +0x14f
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.(*logExporterImp).Start()
      /home/circleci/project/exporter/loadbalancingexporter/log_exporter.go:76 +0x99
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter.TestRollingUpdatesWhenConsumeLogs()
      /home/circleci/project/exporter/loadbalancingexporter/log_exporter_test.go:448 +0x1566
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
==================
--- FAIL: TestRollingUpdatesWhenConsumeLogs (0.09s)
    testing.go:1092: race detected during execution of test
FAIL
coverage: 97.2% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter	0.326s
FAIL
make[2]: *** [../../Makefile.Common:48: do-unit-tests-with-cover] Error 1
make[2]: Leaving directory '/home/circleci/project/exporter/loadbalancingexporter'
make[1]: *** [Makefile:91: for-all] Error 2
make[1]: Leaving directory '/home/circleci/project'
make: *** [Makefile:49: unit-tests-with-cover] Error 2

Exited with code exit status 2

CircleCI received exit code 2
```

--- a/exporter/loadbalancingexporter/log_exporter_test.go
+++ b/exporter/loadbalancingexporter/log_exporter_test.go
@@ -330,6 +330,11 @@ func TestRollingUpdatesWhenConsumeLogs(t *testing.T) {
 	res, err := newDNSResolver(zap.NewNop(), "service-1", "")
 	require.NoError(t, err)
 
+	var lastResolved []string
+	res.onChange(func(s []string) {
+		lastResolved = s
+	})
+
 	resolverCh := make(chan struct{}, 1)
 	counter := 0
 	resolve := [][]net.IPAddr{
@@ -441,7 +446,7 @@ func TestRollingUpdatesWhenConsumeLogs(t *testing.T) {
 	<-consumeCh
 
 	// verify
-	require.Equal(t, []string{"127.0.0.2"}, res.endpoints)
+	require.Equal(t, []string{"127.0.0.2"}, lastResolved)
 	require.Greater(t, atomic.LoadInt64(&counter1), int64(0))
 	require.Greater(t, atomic.LoadInt64(&counter2), int64(0))
 }

--- a/exporter/loadbalancingexporter/trace_exporter_test.go
+++ b/exporter/loadbalancingexporter/trace_exporter_test.go
@@ -327,6 +327,11 @@ func TestRollingUpdatesWhenConsumeTraces(t *testing.T) {
 	res, err := newDNSResolver(zap.NewNop(), "service-1", "")
 	require.NoError(t, err)
 
+	var lastResolved []string
+	res.onChange(func(s []string) {
+		lastResolved = s
+	})
+
 	resolverCh := make(chan struct{}, 1)
 	counter := 0
 	resolve := [][]net.IPAddr{
@@ -437,7 +442,7 @@ func TestRollingUpdatesWhenConsumeTraces(t *testing.T) {
 	<-consumeCh
 
 	// verify
-	require.Equal(t, []string{"127.0.0.2"}, res.endpoints)
+	require.Equal(t, []string{"127.0.0.2"}, lastResolved)
 	require.Greater(t, atomic.LoadInt64(&counter1), int64(0))
 	require.Greater(t, atomic.LoadInt64(&counter2), int64(0))
 }

